tolstoy/src/metadata.rs

```diff
@@ -0,0 +1,96 @@
+// Copyright 2016 Mozilla
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+// Unless required by applicable law or agreed to in writing, software distributed
+// under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR
+// CONDITIONS OF ANY KIND, either express or implied. See the License for the
+// specific language governing permissions and limitations under the License.
+
+use TolstoyResult;
+use uuid;
+use schema;
+use rusqlite;
+use Result;
+
+trait SyncMetadataClientInterface {
+    fn get_remote_head(&self) -> Result<uuid::Uuid>;
+    fn set_remote_head(&mut self, uuid: uuid::Uuid) -> TolstoyResult;
+}
+
+struct SyncMetadataClient {
+    conn: rusqlite::Connection
+}
+
+impl SyncMetadataClient {
+    fn new(conn: rusqlite::Connection) -> Self {
+        SyncMetadataClient {
+            conn: conn
+        }
+    }
+}
+
+impl SyncMetadataClientInterface for SyncMetadataClient {
+    fn get_remote_head(&self) -> Result<uuid::Uuid> {
+        let uuid_query_res = self.conn.query_row(
+            "SELECT value FROM tolstoy_metadata WHERE key = ?",
+            &[&schema::REMOTE_HEAD_KEY], |r| {
+                let raw_uuid: Vec<u8> = r.get(0);
+                match uuid::Uuid::from_bytes(raw_uuid.as_slice()) {
+                    Ok(uuid) => uuid,
+                    Err(e) => panic!("Couldn't parse UUID: {}", e)
```

Comment:
Best not to panic. And this method returns a result! So:

```
let bytes: Vec<u8> = r.get(0);
uuid::Uuid::from_bytes(bytes.as_slice()).map_err(|e| e.into())
```

should suffice to translate the error into your own chain, or perhaps:

```
let bytes: Vec<u8> = r.get(0);
uuid::Uuid::from_bytes(bytes.as_slice())?
```